### DISCUSSION TITLE
How do I remove site collection admin permissions using the SDK?

### DISCUSSION THREAD

**Discussion by plamber** [pnp/pnpcore#1042] [Q&A]
Hello,
I wonder what is the most efficient way to remove the site collection admin permissions to a user using the SDK.

Thank you for the feedback,
Patrick

**Comment by jansenbe** [comment]
@plamber : we don't have a dedicated method for that today, something we indeed should add. 

As a workaround I think using the `GetSiteCollectionAdminsAsync` method to get the current admins, followed by using the `SetSiteCollectionAdminsAsync` method (see https://pnp.github.io/pnpcore/using-the-sdk/admin-sharepoint-security.html#setting-the-administrators-of-a-site-collection) will work as well if you're working with a non-group connected site. 

For group-connected sites you can use Graph APIs to remove the user from the owners group, `PnPContext.Site.GroupId` is automatically populated and will give you the related group for the "common" models, when working with sites linked to shared/private Teams channels getting the groups also involves loading and checking the `ChannelGroupId` and `RelatedGroupId` `ISite` properties.  Once you have the group id use the custom graph API call feature of the SDK (see https://pnp.github.io/pnpcore/using-the-sdk/basics-customapirequests.html#making-a-custom-microsoft-graph-request) to make a request for removing a user from the group owners (see https://learn.microsoft.com/en-us/graph/api/group-delete-owners?view=graph-rest-1.0&tabs=http). You might need to translate the user's UPN to AzureAD id (see https://learn.microsoft.com/en-us/graph/api/user-get?view=graph-rest-1.0&tabs=http).

**Reply by plamber** [reply]
Hi @jansenbe,
For non-group-connected sites using the SetSiteCollectionAdminsAsync will only add new site collection admins but does not delete the existing ones. Is this the expected behavior?

I think I have to get all the site collection admins using the `GetSiteCollectionAdminsAsync` method and update the individual users overwriting the IsSiteAdmin property. What do you think about it? Is anything more efficient than that?

Cheers,
Patrick

**Reply by jansenbe** [reply]
@plamber : I was hoping it would also remove the not provided admins, but if your testing shows that's not the case then some additional work is needed. Currently travelling so a bit timeconstrained, next week I'll have more time to investigate this. `IsSiteAdmin` is not settable, so that will not work. What you can do if you need an urgent solution is do a Fiddler trace when you remove a site collection admin via tenant admin center and then use the custom API feature to do similar from your code. Please share the output here as that will help with building a `RemoveSiteCollectionAdminAsync` method.

**Reply by plamber** [reply]
Hi @jansenbe,
The admin center calls this endpoint `_api/SPOInternalUseOnly.Tenant/SetSiteSecondaryAdministrators` instead of `_api/Microsoft.Online.SharePoint.TenantAdministration.Tenant/SetSiteSecondaryAdministrators`. 

Meet you at the conference :). You will find me at the EasyLife boot.

Cheers

**Reply by plamber** [reply]
Hi @jansenbe,
I found some code that could "demote" a site collection admin without deleting the user from the site collection.

```
var body = new
{
    __metadata = new
    {
        type = "SP.User"
    },
    IsSiteAdmin = false
};
var user = await context.Web.EnsureUserAsync(adminUserPrincipalName);
var accessRequest = new ApiRequest(new HttpMethod("MERGE"), ApiRequestType.SPORest, $"/_api/web/getuserbyid({user.Id})", JsonSerializer.Serialize(body));
await context.Web.ExecuteRequestAsync(accessRequest);
```

The code works only as long there is at least one site collection admin.

Cheers
